File: reference/src/adapters/stubs/jira.ts
/**
 * Stub Jira adapter for testing.
 * Implements IssueTracker interface in-memory.
 */

import type {
  IssueTracker,
  IssueFilter,
  Issue,
  CreateIssueInput,
  UpdateIssueInput,
  IssueEvent,
  EventStream,
} from '../interfaces.js';

export interface StubJiraAdapter extends IssueTracker {
  getIssueCount(): number;
  getStoredIssue(id: string): Issue | undefined;
}

export function createStubJira(): StubJiraAdapter {
  const issues = new Map<string, Issue>();
  const comments = new Map<string, string[]>();
  let nextId = 1;

  return {
    async listIssues(filter: IssueFilter): Promise<Issue[]> {
      let result = Array.from(issues.values());
      if (filter.status) result = result.filter((i) => i.status === filter.status);
      if (filter.labels?.length)
        result = result.filter((i) => filter.labels!.some((l) => i.labels?.includes(l)));
      if (filter.assignee) result = result.filter((i) => i.assignee === filter.assignee);
      if (filter.project) result = result.filter((i) => i.url.includes(filter.project!));
      return result;
    },

    async getIssue(id: string): Promise<Issue> {
      const issue = issues.get(id);
      if (!issue) throw new Error(`Issue "${id}" not found`);
      return issue;
    },

    async createIssue(input: CreateIssueInput): Promise<Issue> {
      const id = `JIRA-${nextId++}`;
      const issue: Issue = {
        id,
        title: input.title,
        description: input.description,
        status: 'open',
        labels: input.labels,
        assignee: input.assignee,
        url: `https://jira.example.com/browse/${id}`,
      };
      issues.set(id, issue);
      return issue;
    },

    async updateIssue(id: string, input: UpdateIssueInput): Promise<Issue> {
      const issue = issues.get(id);
      if (!issue) throw new Error(`Issue "${id}" not found`);
      if (input.title !== undefined) issue.title = input.title;
      if (input.description !== undefined) issue.description = input.description;
      if (input.labels !== undefined) issue.labels = input.labels;
      if (input.assignee !== undefined) issue.assignee = input.assignee;
      return issue;
    },

    async transitionIssue(id: string, transition: string): Promise<Issue> {
      const issue = issues.get(id);
      if (!issue) throw new Error(`Issue "${id}" not found`);
      issue.status = transition;
      return issue;
    },

    async addComment(id: string, body: string): Promise<void> {
      if (!issues.has(id)) throw new Error(`Issue "${id}" not found`);
      const existing = comments.get(id) ?? [];
      existing.push(body);
      comments.set(id, existing);
    },

    async getComments(id: string): Promise<Array<{ body: string }>> {
      return (comments.get(id) ?? []).map((body) => ({ body }));
    },

    watchIssues(_filter: IssueFilter): EventStream<IssueEvent> {
      return {
        async *[Symbol.asyncIterator]() {
          // Stub: no events
        },
      };
    },

    getIssueCount(): number {
      return issues.size;
    },

    getStoredIssue(id: string): Issue | undefined {
      return issues.get(id);
    },
  };
}

File: sdk-python/src/ai_sdlc/adapters/stubs/jira.py
"""Stub Jira adapter for testing."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import AsyncIterator

from ai_sdlc.adapters.interfaces import (
    CreateIssueInput,
    Issue,
    IssueComment,
    IssueEvent,
    IssueFilter,
    UpdateIssueInput,
)


class StubJiraAdapter:
    def __init__(self) -> None:
        self._issues: dict[str, Issue] = {}
        self._comments: dict[str, list[str]] = {}
        self._next_id = 1

    async def list_issues(self, filter: IssueFilter) -> list[Issue]:
        result = list(self._issues.values())
        if filter.status:
            result = [i for i in result if i.status == filter.status]
        if filter.labels:
            result = [
                i for i in result
                if i.labels and any(lbl in i.labels for lbl in filter.labels)
            ]
        if filter.assignee:
            result = [i for i in result if i.assignee == filter.assignee]
        return result

    async def get_issue(self, id: str) -> Issue:
        if id not in self._issues:
            raise KeyError(f'Issue "{id}" not found')
        return self._issues[id]

    async def create_issue(self, input: CreateIssueInput) -> Issue:
        iid = f"JIRA-{self._next_id}"
        self._next_id += 1
        issue = Issue(
            id=iid, title=input.title, status="open",
            url=f"https://jira.example.com/browse/{iid}",
            description=input.description,
            labels=input.labels, assignee=input.assignee,
        )
        self._issues[iid] = issue
        return issue

    async def update_issue(self, id: str, input: UpdateIssueInput) -> Issue:
        old = self._issues.get(id)
        if not old:
            raise KeyError(f'Issue "{id}" not found')
        self._issues[id] = Issue(
            id=old.id,
            title=input.title if input.title is not None else old.title,
            status=old.status,
            url=old.url,
            description=(
                input.description
                if input.description is not None
                else old.description
            ),
            labels=input.labels if input.labels is not None else old.labels,
            assignee=(
                input.assignee if input.assignee is not None else old.assignee
            ),
        )
        return self._issues[id]

    async def transition_issue(self, id: str, transition: str) -> Issue:
        old = self._issues.get(id)
        if not old:
            raise KeyError(f'Issue "{id}" not found')
        self._issues[id] = Issue(
            id=old.id, title=old.title, status=transition,
            url=old.url, description=old.description,
            labels=old.labels, assignee=old.assignee,
        )
        return self._issues[id]

    async def add_comment(self, id: str, body: str) -> None:
        if id not in self._issues:
            raise KeyError(f'Issue "{id}" not found')
        self._comments.setdefault(id, []).append(body)

    async def get_comments(self, id: str) -> list[IssueComment]:
        return [IssueComment(body=b) for b in self._comments.get(id, [])]

    def watch_issues(
        self, filter: IssueFilter,
    ) -> AsyncIterator[IssueEvent]:
        async def _empty() -> AsyncIterator[IssueEvent]:
            return
            yield  # noqa: RET504  # pragma: no cover
        return _empty()

    def get_issue_count(self) -> int:
        return len(self._issues)


def create_stub_jira() -> StubJiraAdapter:
    return StubJiraAdapter()
